### PR TITLE
IGNITE-15766 CPP thin: Implement continuous query

### PR DESCRIPTION
**Pull request by isapego**
- Implemented Continuous Queries for C++ thin client:
- Implemented `ThreadPool` class.
- Added handling of server notifications in `userThreadPool`
- Remote filters are not implemented yet (IGNITE-16291).

### The Contribution Checklist
- [x] There is a single JIRA ticket related to the pull request. 
- [x] The web-link to the pull request is attached to the JIRA ticket.
- [x] The JIRA ticket has the _Patch Available_ state.
- [x] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [x] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [x] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [x] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/platforms/cpp/common/src/common/thread_pool.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        {
+            Stop();
+        }
+
+        void ThreadPool::Start()
+        {
+            if (started)
+                return;
+
+            started = true;
```
Comment (ptupitsyn) [inline modules/platforms/cpp/common/src/common/thread_pool.cpp:51]: Do we care about race condition here?

File: modules/platforms/cpp/thin-client-test/src/continuous_query_test.cpp [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+    typedef impl::cache::query::continuous::ContinuousQueryImpl<int, TestEntry> QueryImplType;
+    
+    BOOST_CHECK_EQUAL(static_cast<int>(QueryImplType::DEFAULT_TIME_INTERVAL),
+        static_cast<int>(QueryType::DEFAULT_TIME_INTERVAL));
+
+    BOOST_CHECK_EQUAL(static_cast<int>(QueryImplType::DEFAULT_BUFFER_SIZE),
+        static_cast<int>(QueryType::DEFAULT_BUFFER_SIZE));
+}
+
+BOOST_AUTO_TEST_SUITE_END()
```
Comment (ptupitsyn) [inline modules/platforms/cpp/thin-client-test/src/continuous_query_test.cpp:375]: Let's add a test to receive `Expired` events when `IncludeExpired` is true.

File: modules/platforms/cpp/thin-client/include/ignite/impl/thin/cache/query/query_fields_cursor_impl.h [hunk 3]
```diff
@@ -1,147 +0,0 @@
-/*
```
Comment (ptupitsyn) [inline modules/platforms/cpp/thin-client/include/ignite/impl/thin/cache/query/query_fields_cursor_impl.h:1]: Was this file deleted intentionally?

File: modules/platforms/cpp/thin-client/include/ignite/thin/cache/event/cache_entry_event_listener.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    {
+        namespace cache
+        {
+            namespace event
+            {
+                /**
+                 * Cache entry event listener.
+                 */
+                template<typename K, typename V>
+                class CacheEntryEventListener
```
Comment (ptupitsyn) [inline modules/platforms/cpp/thin-client/include/ignite/thin/cache/event/cache_entry_event_listener.h:42]: Can we provide `OnDisconnected` here as well, so the user knows that continuous query has stopped due to connection loss? Or is there some other API to deal with this situation?

File: modules/platforms/cpp/common/src/common/thread_pool.cpp (hunk 1 above)
Comment (isapego) [reply modules/platforms/cpp/common/src/common/thread_pool.cpp:51]: Start is not supposed to be called from several threads at the same time.

File: modules/platforms/cpp/common/src/common/thread_pool.cpp [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+            for (std::vector<SP_WorkerThread>::iterator it = threads.begin(); it != threads.end(); ++it)
+                it->Get()->Start();
+        }
+
+        void ThreadPool::Stop()
+        {
+            if (stopped)
+                return;
+
+            stopped = true;
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/common/src/common/thread_pool.cpp:62]: Well, just as above -- too many things may go wrong :) This is just something that one should not be doing.

File: modules/platforms/cpp/common/src/common/thread_pool.cpp [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+                if (!task.IsValid())
+                    break;
+
+                try
+                {
+                    task.Get()->Execute();
+                }
+                catch (...)
+                {
+                    // No-op.
```
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/common/src/common/thread_pool.cpp:151]: I agree, but there is not yet logging in C++ thin client. But I agree that we should handle this differently.

File: modules/platforms/cpp/thin-client-test/src/continuous_query_test.cpp (hunk 2 above)
Comment (isapego) [reply modules/platforms/cpp/thin-client-test/src/continuous_query_test.cpp:375]: Agree

File: modules/platforms/cpp/thin-client/include/ignite/impl/thin/cache/query/query_fields_cursor_impl.h (hunk 3 above)
Comment (isapego) [reply modules/platforms/cpp/thin-client/include/ignite/impl/thin/cache/query/query_fields_cursor_impl.h:1]: Yeah, it was just some old unused file that was just lying there. Was not used anywhere.

File: modules/platforms/cpp/thin-client/include/ignite/thin/cache/event/cache_entry_event_listener.h (hunk 4 above)
Comment (isapego) [reply modules/platforms/cpp/thin-client/include/ignite/thin/cache/event/cache_entry_event_listener.h:42]: That's a really good idea, I've somehow missed this part. Need to add a test too.

File: modules/platforms/cpp/common/src/common/thread_pool.cpp [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+            }
+
+            SP_ThreadPoolTask res = tasks.front();
+            tasks.pop_front();
+            return res;
+        }
+
+        void ThreadPool::TaskQueue::Unblock()
+        {
+            concurrent::CsLockGuard guard(mutex);
```
Comment (ivandasch) [inline modules/platforms/cpp/common/src/common/thread_pool.cpp:119]: Igor, is it ok that tasks when thread-pool is stopped are abandoned?
May be it is good idea to complete task exceptionally?

File: modules/platforms/cpp/common/src/common/thread_pool.cpp [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+                return;
+
+            concurrent::CsLockGuard guard(mutex);
+
+            tasks.push_back(task);
+
+            waitPoint.NotifyOne();
+        }
+
+        SP_ThreadPoolTask ThreadPool::TaskQueue::Pull()
```
Comment (ivandasch) [inline modules/platforms/cpp/common/src/common/thread_pool.cpp:98]: A little bit strange naming for queue, may be Pop is better?

File: modules/platforms/cpp/common/src/common/thread_pool.cpp (hunk 1 above)
Comment (ivandasch) [reply modules/platforms/cpp/common/src/common/thread_pool.cpp:51]: Agree with Pavel, may be better to add mutex?

Comment (isapego) [reply modules/platforms/cpp/common/src/common/thread_pool.cpp:51]: Guys, it's only called on client start. Client start can not be called concurrently on the same instance as user only gets an instance of `IgniteClient` once client is already started. If they call `IgniteClient::Start()` from two different threads they will simply get 2 separate instances of client.

File: modules/platforms/cpp/common/src/common/thread_pool.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (isapego) [reply modules/platforms/cpp/common/src/common/thread_pool.cpp:62]: The same here - it is only stopped in destructor, which can not be called from different threads.

File: modules/platforms/cpp/common/src/common/thread_pool.cpp (hunk 8 above)
Comment (isapego) [reply modules/platforms/cpp/common/src/common/thread_pool.cpp:98]: Agree.

File: modules/platforms/cpp/common/src/common/thread_pool.cpp (hunk 7 above)
Comment (isapego) [reply modules/platforms/cpp/common/src/common/thread_pool.cpp:119]: Will handle when will fix issue that @ptupitsyn suggested.

**Review by ptupitsyn — APPROVED** [review]

**Review by ivandasch — APPROVED** [review]

**Pull request closed without merging**